packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -758,31 +763,33 @@ function areCompatibleNodes(client: Node, ssr: Node, vnode: VNode, renderer: Ren
         return false;
     }
 
-    const clientAttrsNames: string[] = getProperty(client, 'getAttributeNames').call(client);
+    if (isUndefined(vnode.parts)) {
```

Comment:
Why do we only validate these attributes when `vnode.parts` is undefined?

Reply:
@nolanlawson I ended up [changing this](https://github.com/salesforce/lwc/pull/4071/commits/c3dddb4565966b351413c05bf1596a8d4c99a953#diff-a9673af211cdbdb5096c7ee4533b234a8bf90e9271341048b42ffcd6c0e143ceR769-R774), my original assumption was incorrect.

This validation checks the attributes of the root static element match and remounts if it doesn't.

The issue is when there are expressions on the root element, the validation will fail because the fragment and SSR-generated element don't match.

The SSR element will have the attribute expressions applied while the fragment doesn't because the patch routines haven't been called yet.

EX: 
```html
<div class={dynamic}></div>
```

```js
export default class extends LightningElement {
    dynamic = 'foo';
}
```

If this component has scoped styles it will render:

```html
<div class="scoped_token foo"></div>
// SSR
```

```html
<div class="scoped_token"></div>
// hydration
```

I've added an extra check when validation fails here to defer validation to `haveCompatibleStaticParts` if there's a static part referencing the root element.